### ISSUE TITLE
Broken LLVM module when with @call and BoundFns

### ISSUE THREAD

**Issue by adrusi** [ziglang/zig#5695]
```zig
const std = @import("std");

const T = struct {
    x: u8,

    fn f(self: *const @This(), y: u8) u8 {
        return self.x + y;
    }
};

fn a(comptime f: var, args: var) u8 {
    return @call(.{}, f, args);
}

pub fn main() void {
    var t = T { .x = 5 };
    var fr = a(t.f, .{ 7 });
    std.debug.warn("{}\n", .{fr});
}
```

```
broken LLVM module found: Instruction does not dominate all uses!

  %t = alloca %T, align 1

  %0 = call fastcc i8 @T.f(%T* %t, i8 7), !dbg !16621


This is a bug in the Zig compiler.

Unable to dump stack trace: debug info stripped
```

There's also this apparently different error:

```zig
const std = @import("std");

var frame: anyframe = undefined;

const T = struct {
    x: u8,

    fn f(self: *const @This(), y: u8) u8 {
        suspend { frame = @frame(); }
        return self.x + y;
    }
};

fn U(comptime A: type) type {
    return struct {
        t: A,

        fn g(self: *@This()) u8 {
            return a(self.t.f, .{ 7 });
        }
    };
}

fn a(comptime f: var, args: var) u8 {
    return await @call(.{ .modifier = .async_kw }, f, args);
    // return @call(.{}, f, args);
}

pub fn main() void {
    var u = U(T) { .t = .{ .x = 5 } };
    var fr = async u.g();
    resume frame;
    std.debug.warn("{}\n", .{nosuspend await fr});
}
```

```
broken LLVM module found: !dbg attachment points at wrong subprogram for function

!17472 = distinct !DISubprogram(name: "a", scope: !15333, file: !15333, line: 24, type: !17473, scopeLine: 24, flags: DIFlagStaticMember, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition, unit: !2, retainedNodes: !1087)

void (%"@Frame(a)"*, i64)* @a

  %11 = load %T*, %T** %3, align 8, !dbg !16685

!16685 = !DILocation(line: 19, column: 26, scope: !16686)

!16686 = distinct !DILexicalBlock(scope: !16677, file: !15333, line: 18, column: 33)

!16677 = distinct !DISubprogram(name: "U(T).g", scope: !15337, file: !15333, line: 18, type: !16678, scopeLine: 18, flags: DIFlagStaticMember, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition, unit: !2, retainedNodes: !16681)

Instruction does not dominate all uses!

  %3 = getelementptr inbounds %"@Frame(U(T).g)", %"@Frame(U(T).g)"* %0, i32 0, i32 10, !dbg !16683

  %11 = load %T*, %T** %3, align 8, !dbg !16685


This is a bug in the Zig compiler.

Unable to dump stack trace: debug info stripped
```

**Comment by codehz** [comment]
another minimal poc: 
```zig
const std = @import("std");

const TestStruct = struct {
    val: u32, // IMPORTANT
    fn testFunction(self: *@This()) void {}
};

fn anotherFunction(f: anytype) void {
    f();
}

pub fn main() anyerror!void {
    var r: TestStruct = .{ .val = 5 };
    anotherFunction(r.testFunction);
}
```

it works if remove the val field

**Comment by remorses** [comment]
Will this be fixed by #9484?
